enzyme/Enzyme/MLIR/Interfaces/EnzymeLogicReverse.cpp

```diff
@@ -0,0 +1,190 @@
+  if (oBB->hasNoPredecessors()){
+    SmallVector<mlir::Value, 2> retargs;
+    for (Value attribute : gutils->oldFunc.getFunctionBody().getArguments()) {
+      Value attributeGradient = gutils->invertPointerM(attribute, revBuilder);
+      retargs.push_back(attributeGradient);
+    }
+    revBuilder.create<func::ReturnOp>(oBB->rbegin()->getLoc(), retargs);
+  }
+  else {
+    SmallVector<Block *> blocks;
+    SmallVector<APInt> indices;
+    SmallVector<ValueRange> arguments;
+    ValueRange defaultArguments;
+    Block * defaultBlock;
+    int i = 1;
+    for (Block * predecessor : oBB->getPredecessors()){
+      Block * predecessorRevMode = gutils->mapReverseModeBlocks.lookupOrNull(predecessor);
+
+      SmallVector<Value> operands;
+      auto argumentsIt = gutils->mapBlockArguments.find(predecessor);
+      if (argumentsIt != gutils->mapBlockArguments.end()){
+        for(auto operandOld : argumentsIt->second){
+          if (oBB == operandOld.first.getParentBlock() && gutils->hasInvertPointer(operandOld.first)){
+            operands.push_back(gutils->invertPointerM(operandOld.first, revBuilder));
+          }
+          else{
+            if (auto iface = operandOld.first.getType().cast<AutoDiffTypeInterface>()) {
+              Value nullValue = iface.createNullValue(revBuilder, oBB->rbegin()->getLoc());
+              operands.push_back(nullValue);
+            }
+            else{
+              llvm_unreachable("non canonial null value found");
+            }
+          }
+        }
+      }
+
+      if (predecessor != *(oBB->getPredecessors().begin())){
+        blocks.push_back(predecessorRevMode);
+        indices.push_back(APInt(32, i++));
+        arguments.push_back(ValueRange(operands));
+      }
+      else{
+        defaultBlock = predecessorRevMode;
+        defaultArguments = ValueRange(operands);
+      }
+    }
+    //Remove Dependency to CF dialect
+    if (std::next(oBB->getPredecessors().begin()) == oBB->getPredecessors().end()){
+      //If there is only one block we can directly create a branch for simplicity sake
+      revBuilder.create<cf::BranchOp>(gutils->getNewFromOriginal(&*(oBB->rbegin()))->getLoc(), defaultBlock, defaultArguments);
+    }
+    else{
+      Value cache = gutils->insertInitBackwardCache(gutils->getIndexCacheType());
+      Value flag = revBuilder.create<enzyme::PopCacheOp>(oBB->rbegin()->getLoc(), gutils->getIndexType(), cache);
+
+      revBuilder.create<cf::SwitchOp>(oBB->rbegin()->getLoc(), flag, defaultBlock, defaultArguments, ArrayRef<APInt>(indices), ArrayRef<Block *>(blocks), ArrayRef<ValueRange>(arguments));
+      
+      int j = 0;
+      for (Block * predecessor : oBB->getPredecessors()){
+        Block * newPredecessor = gutils->getNewFromOriginal(predecessor);
+        OpBuilder predecessorBuilder(newPredecessor, std::prev(newPredecessor->end()));
+
+        Value indicator = predecessorBuilder.create<arith::ConstantIntOp>(oBB->rbegin()->getLoc(), j++, 32);
```

Comment:
Constants are a bit unfortunate dialect dependency-wise, but we may handle this similarly to `cf::`. Please add a TODO. An alternative would be to have an `enzyme.constant`. Or, in this particular, case, we don't actually need the SSA value for `indicator`, but we can rather have `PushCashOp` use the attribute instead.